src/rust/engine/src/rule_graph.rs

```diff
@@ -451,12 +451,20 @@ impl<'t> GraphMaker<'t> {
         // If no candidates were fulfillable, this rule is not fulfillable.
         unfulfillable_diagnostics.push(Diagnostic {
           params: params.clone(),
-          reason: format!(
-            "no rule was available to compute {} with parameter type{} {}",
-            type_constraint_str(product),
-            if params.len() > 1 { "s" } else { "" },
-            params_str(&params),
-          ),
+          reason: if params_str(&params) == "()" {
```

Comment:
Rather than looking at the string value here, you should be able to check `params.is_empty()`.